---
title: Explore UniProtKB entry
categories: Get_started, UniProtKB
---
The UniProtKB database contains individual protein entries that provide detailed information about the protein. At the top of the page, you can see the UniProt entry accession and name, the name of the protein and gene, the organism, whether the protein entry has been manually reviewed by a UniProt curator, its annotation score, and the evidence level for its existence.

You can click here to run a BLAST sequence similarity search on the entry, align it with all its isoforms, download the entry in various formats, or add it to the basket to save for later.

![](https://github.com/ebi-uniprot/uniprot-manual/blob/main/images/Explore%20UniProtKB%20entry_1.png?raw=true)

Information in a UniProtKB entry is structured into a number of sections to help you find the data that you are looking for. The tabs allow you to switch between the entry, a graphical view of sequence features (Feature viewer), publications, and external links.

The navigation bar on the left-hand side of the screen lists the entry sections. They are shortcuts that help you jump to specific sections, allowing you to move easily within the entry. The navigation bar stays in the same place on the screen as you move up and down in an entry so that you can quickly navigate to sections of interest. 

![](https://github.com/ebi-uniprot/uniprot-manual/blob/main/images/Explore%20UniProtKB%20entry_2.png?raw=true)

You may see evidence tags in the content, providing details on the source of this information. 

![](https://github.com/ebi-uniprot/uniprot-manual/blob/main/images/Explore%20UniProtKB%20entry_3.png?raw=true)

The feature viewer allows you to see all sequence features together in a visual manner. Features are arranged into categories such as domains and sites, molecule processing, post-translational modifications, mutagenesis, etc. You can click on a category to expand it further. 

The ruler on top represents the sequence length of the protein. To see more information about the feature, you can click on it. This will bring up the tooltip with information and also highlight the sequence position of the feature.

Letâ€™s click on this active site feature. To see which other features are colocalized with this active site, let us zoom in further. This can be done by dragging the handles on the ruler.

![](https://github.com/ebi-uniprot/uniprot-manual/blob/main/images/Explore%20UniProtKB%20entry_4.png?raw=true)

We can now see that our feature coincides with the peak in the Variants, so we click on the Variants to expand the view.

![](https://github.com/ebi-uniprot/uniprot-manual/blob/main/images/Explore%20UniProtKB%20entry_5a.png?raw=true)

We find Variants colour-coded to indicate the involvement with the disease. You can click on them to see more details including the evidence source. The Variants view maps all variants on a matrix with the sequence position on the x axis and the possible amino acid changes on the y axis. You can filter the variants to see only UniProt reviewed data using filters on the left hand-side. 

For more help on how to use UniProt:
* visit our [Help center](https://www.uniprot.org/help)
* watch our [video tutorials](https://www.youtube.com/c/uniprotvideos/videos)
* sign up for [courses](https://www.ebi.ac.uk/training/search-results?query=uniprot&domain=ebiweb_training&page=1&facets=)
* or [contact us](https://www.uniprot.org/contact) directly.