---
title: Where does UniProt data come from?
categories: Get_started,faq
---

UniProt provides both sequence data and associated functional information, derived from a range of sources.

More than 95% of the protein sequences provided by UniProtKB come from the translations of coding sequences (CDS) submitted to the ENA/GenBank/DDBJ nucleotide sequence resources of the International Nucleotide Sequence Database Collaboration (INSDC).
These CDS are either generated by gene prediction programs or are experimentally proven. The translated CDS sequences are automatically transferred to the TrEMBL section of UniProtKB. The TrEMBL records can be selected for further manual annotation and then integrated into the UniProtKB/Swiss-Prot section.

In addition to translated CDS, UniProtKB protein sequences may come from:
* The PDB database
* Sequences experimentally obtained by direct protein sequencing and submitted to UniProt
* Sequences scanned from the literature
* Sequences derived from gene prediction but which have not been submitted to ENA/GenBank/DDBJ. These are imported from resources such as Ensembl and RefSeq

Importing and combining sequences from a range of sources means that UniProt provides a complete collection of protein sequences and contributes to consistency of protein sets across various sequence resources.

UniParc is designed to capture all publicly available protein sequence data and contains all the protein sequences from the main publicly available protein sequence databases. A complete [list of the source databases](https://beta.uniprot.org/help/uniparc) is available on the UniProt website.

UniRef provides clustered sets of sequences from UniProtKB and selected UniParc records.

Functional information is found in the UniProt Knowledgebase. UniProt attempts to attach as much functional information as possible to each protein sequence to provide users with an overview of the available information for a given protein. This information is added manually by the UniProt biocurators who are all trained biologists or added automatically through various annotation systems which have been developed within the group.

Manual curation consists of a critical review of experimental and predicted data for each protein as well as manual verification of each protein sequence.
Curation methods applied include:
* manual extraction and structuring of information from the literature
* manual verification of results from computational analyses
* mining and integration of large-scale data sets
* continuous updating as new information becomes available

You can find more information about the [manual curation](https://beta.uniprot.org/help/manual_curation) process on the UniProt website.

UniProt has developed two prediction systems to automatically annotate UniProtKB/TrEMBL in a scalable manner with a high degree of accuracy:
1. UniRule is a collection of manually curated annotation rules which define annotations that can be propagated based on specific conditions
2. The Statistical Automatic Annotation System (SAAS) is an automatic decision-tree based rule-generating system

You can find out more information about the [automatic annotation pipeline](https://beta.uniprot.org/help/automatic_annotation) on the UniProt website.

For more help on how to use UniProt:
* visit our [Help center](https://beta.uniprot.org/help)
* watch our [video tutorials](https://www.youtube.com/c/uniprotvideos/videos)
* sign up for [courses](https://www.ebi.ac.uk/training/search-results?query=uniprot&domain=ebiweb_training&page=1&facets=)
* or [contact us](https://www.uniprot.org/contact) directly.